File: flow/src/main/java/it/unibo/coordination/flow/Activity.java
package it.unibo.coordination.flow;

import java.util.Optional;

public interface Activity<A, C, R> {
    Continuation<C> onBegin(A argument) throws Exception;

    Continuation<C> onStep(C context) throws Exception;

    R onEnd(C context, Optional<Exception> exception) throws Exception;

    default Continuation<C> onError(C context, Exception exception) {
        return Continuation.stop(context);
    }
}

File: control/src/main/kotlin/it/unibo/coordination/control/Activity.kt
package it.unibo.coordination.control

interface Activity<E, T, R> {

    interface Controller<E, T, R> {
        fun stop(result: R)
        fun restart(input: E)
        fun pause(data: T)
        fun pause()
        fun `continue`(data: T)
        fun `continue`()
        fun resume()
    }

    fun onBegin(input: E, controller: Controller<E, T, R>)
    fun onStep(input: E, lastData: T, controller: Controller<E, T, R>)
    fun onEnd(input: E, lastData: T, result: R, controller: Controller<E, T, R>)
}